graphql_compiler/compiler/ir_lowering_match/utils.py

```diff
@@ -219,3 +220,104 @@ def construct_where_filter_predicate(simple_optional_root_info):
 # A CompoundMatchQuery is a representation of several MatchQuery objects containing
 #   - match_queries: a list MatchQuery objects
 CompoundMatchQuery = namedtuple('CompoundMatchQuery', ('match_queries'))
+
+
+class OptionalTraversalTrie(dict):
```

Comment:
Let's discuss this in person, and figure out whether and how it might relate to the QueryMetadataTable work that has been happening in parallel.